### ISSUE TITLE
Add support for Extension Auto-Installer

### ISSUE THREAD

**Issue by blackwind** [hbenl/vscode-firefox-debug#21]
[Extension Auto-Installer](https://addons.mozilla.org/en-US/firefox/addon/autoinstaller/) support would give add-on developers the power to test their changes without restarting Firefox. `jpm` officially supports this via its `watchpost` command, but it sounds like it's as simple as POSTing a new .xpi, so WebExtensions developers could benefit from this as well.

**Comment by hbenl** [comment]
@blackwind @mbarzeev I have just published a new version that supports this. Just add `"reAttach": true` to your launch configuration and the Extension Auto-Installer will be automatically added to the Firefox debugging profile and used to (re-)install your add-on every time you start a debugging session. I have tested this with jetpack add-ons and WebExtensions. It should also work with restartless legacy add-ons but probably not with XUL overlays (although I haven't tested this).

**Comment by blackwind** [comment]
Well, it appears to work, though I initially thought it was broken because it takes a really, really, REALLY long time to launch Firefox:

* Without `reAttach`: 5 seconds
* With `reAttach`: 36 seconds

This is on a near-top-of-the-line PC. Is there anything you can do about that?

**Comment by hbenl** [comment]
Oops, that was a silly mistake I made there. It's fixed now.

**Comment by blackwind** [comment]
Confirmed. Cheers!

**Comment by blackwind** [comment]
Would it be possible to immediately submit changes to Firefox when a file is saved without restarting the VSCode debug session?

**Comment by hbenl** [comment]
Yes, that would be possible, but I'd still have to create the `xpi` file using `jpm` and that's what's taking most of the time, restarting the debug session is pretty fast compared to that. So the performance improvement would probably be barely noticeable. All you would save would be a mouse click (or key combination).

**Comment by hbenl** [comment]
I am looking at providing such a feature for WebExtensions using the extension proxy file feature described [here](https://developer.mozilla.org/en-US/Add-ons/Setting_up_extension_development_environment). But AFAIK that feature won't work with jetpack add-ons because they have to be processed by `jpm`.

**Comment by blackwind** [comment]
This proxy feature doesn't appear to be WebExtensions-exclusive. Remember that all `jpm` is doing is creating an `install.rdf` based on `package.json`, adding `bootstrap.js` (which is always the same file), zipping everything up, and giving it a `.xpi` extension. Once you have `install.rdf` from the initial run of `jpm xpi`, you can add files to the zip manually or unzip it and use the proxy feature.

**Comment by hbenl** [comment]
Hmm... yes... this sounds like it could work! Not exactly trivial but I'll give it a try (though it might take some time). Thanks for the idea!

**Comment by hbenl** [comment]
The latest version supports add-on reloading:
It's not completely automatic, you'll have to execute the "Firefox: Reload add-on" command from the VS Code command palette - I recommend to create a keyboard shortcut for this command.

**Comment by blackwind** [comment]
Appears to work, but having to use two keys is a bit of a frustrating user experience. Is there an API limitation that necessitated this design?

**Comment by hbenl** [comment]
This was just the first step, automatic reloading comes next.

**Comment by hbenl** [comment]
You can now use the `reloadOnChange` configuration property to automatically reload your add-on.

**Comment by blackwind** [comment]
Unable to test, as the current version is having trouble running Firefox. It sits there for a while without launching `firefox.exe`, and then throws:

> Error: connect ECONNREFUSED 127.0.0.1:6000

**Comment by hbenl** [comment]
It works for me on Linux and Windows 7. What's your OS version and VS Code version? And could you post your launch.json?

**Comment by hbenl** [comment]
Couldn't reproduce on Windows 10 either.

**Comment by blackwind** [comment]
Windows 10 1703, VS Code 1.12.1.

```
{
  "version": "0.2.0",
  "configurations": [
    {
      "name": "Run",
      "type": "firefox",
      "request": "launch",
      "addonType": "addonSdk",
      "addonPath": "${workspaceRoot}",
      "reAttach": true,
      "reloadOnChange": {
        "watch": "${workspaceRoot}/**/*.js",
        "ignore": "**/node_modules/**"
      },
      "skipFiles": ["resource://**/sdk/**", "webpack://**"]
    },
  ]
}
```

**Comment by hbenl** [comment]
Well, I have a slightly different Windows version (1607), but I don't think that's the problem.
Could you try
- without `reAttach` or
- without `reloadOnChange` or
- with a non-addon configuration

and see if any of those work?

**Comment by blackwind** [comment]
Same thing in all three scenarios.

**Comment by hbenl** [comment]
Hmmm... if you want, you could try debugging the debug adapter to try to find out where it fails:
* clone this repository
* open the folder in VS Code
* run `npm install` in VS Code's integrated terminal
* run the build task in VS Code's command palette
* start the `debug server` launch configuration
* start the `web test` launch configuration (while `debug server` is still running)

If you can reproduce the problem (i.e. Firefox doesn't start) you could then try stepping through the `launchFirefox` method in `src/util/launcher.ts`...

**Comment by blackwind** [comment]
That's a bit too much for me to get into right now, but I just put two and two together, and I think it's the Creators Update at fault. I've noticed clicking on links in my email client no longer does anything since I upgraded to 1703, and I just tested clicking on links in several other programs that were previously working and found the same behavior. I switched my default browser to Edge, and sure enough, each program was able to open the link in Edge; when I switched back to Firefox, nada. Something fishy's going on here.
